File: 1650-lowest-common-ancestor-of-a-binary-tree-iii/1650-lowest-common-ancestor-of-a-binary-tree-iii.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node parent;
};
*/

class Solution {
    public Node lowestCommonAncestor(Node p, Node q) {
        Set<Node> set = new HashSet<>();
        while(true) {
            if(p != null && !set.add(p)) return p;
            if(q != null && !set.add(q)) return q;
            if(p != null) p = p.parent;
            if(q != null) q = q.parent;
        }
    }
}

File: 1650-lowest-common-ancestor-of-a-binary-tree-iii/1650-lowest-common-ancestor-of-a-binary-tree-iii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None
"""

class Solution:
    def lowestCommonAncestor(self, p: 'Node', q: 'Node') -> 'Node':
        first, second = p, q
        while first != second:
            first = first.parent if first else q
            second = second.parent if second else p
        return first